securedrop/journalist.py

```diff
@@ -593,11 +602,24 @@ def reply():
                         [crypto_util.getkey(g.sid), config.JOURNALIST_KEY],
                         output=store.path(g.sid, filename))
     reply = Reply(g.user, g.source, filename)
-    db_session.add(reply)
-    db_session.commit()
 
-    flash("Thanks! Your reply has been stored.", "notification")
-    return redirect(url_for('col', sid=g.sid))
+    try:
+        db_session.add(reply)
+        db_session.commit()
+    except Exception as exc:
+        flash("An unexpected error occurred! Please check the application "
+              "logs or inform your adminstrator.", "error")
+        # We take a cautious approach to logging here because we're dealing
+        # with responses to sources. It's possible the exception message could
+        # contain information we don't want to write to disk.
+        app.logger.error(
+            "Reply from '{}' (id {}) failed: {}!".format(g.user.username,
+                                                         g.user.id,
+                                                         exc.__class__))
+    else:
+        flash("Thanks! Your reply has been stored.", "notification")
+    finally:
+        return redirect(url_for('col', sid=g.sid))
```

Comment:
As we just discussed f2f, this 404 is occurring in the `finally` block, and any case where the source is deleted won't be gracefully handled. To handle this situation one could delete this `finally` block and instead:

1. Redirect to the home page if the source is deleted so that a flashed message is displayed and no 404 occurs. Nice advantage: one could also then flash a message specific to this situation ("Sorry, the source was deleted by another user."). You can simply write a unit test for the first case just by taking one of the tests below and adding a line where you delete the source.
2. Else you redirect to the `url_for('col', sid=g.sid)` and flash the message as you are currently doing. And you've already got the unit tests for this written :+1: